Fast RB for QM

stavros11 opened the pull request:
Following the discussion in https://github.com/qiboteam/qibocal/pull/951#issuecomment-2258107510, I am opening a draft PR for this as others may be interested. 

As described in that comment, it provides a faster (close to ideal time) implementation of single qubit RB for QM, following the example from https://github.com/qua-platform/qua-libs/blob/main/Quantum-Control-Applications/Superconducting/Single-Fixed-Transmon/16a_randomized_benchmarking.py with the sampling of random sequences happenning on the device. It also provides the possibility to disable the last inverse gate and apply filtered post-processing, or reduce the relaxation time to execute restless.

Example runcard:
```yaml
platform: qw11q

targets: ["D1"]  # only works for one qubit (no two qubits in parallel)

actions:

  - id: rb_qua
    operation: rb_ondevice
    parameters:
      num_of_sequences: 10000  # Number of random sequences.
      max_circuit_depth: 500  # Maximum circuit depth.
      delta_clifford: 10  # Depth step. Depths applied are [1, delta, 2 * delta, ..., max].
      n_avg: 1  # Number of repetitions of the same circuit (for averaging).
      save_sequences: True  # Download the Clifford gate indices used from the device.
      apply_inverse: True  # Apply inverse gate at the end of every circuit. If `False` it will apply "filtered" post-processing.
```

Example report: http://login.qrccluster.com:9000/jKoOT3NjREyjnnb1m-R_0A==/. More examples here: https://github.com/qiboteam/qibolab_platforms_qrc/pull/154#issuecomment-2248208893.

The are no plans to merge this soon as the code quality is not very good, so I would not spend time on reviews at this stage. Any suggestions for improvements are welcome though, for example for the data saving that is a bit messy.

stavros11 committed:
feat: first commit of rb using qua

stavros11 committed:
acquisition working

stavros11 committed:
acquisition working

stavros11 committed:
acquisition working

stavros11 committed:
remove specialized data save and load

stavros11 committed:
fix fitting

stavros11 committed:
fix data propagation

stavros11 committed:
fix: use platfrom values for sweetspots and device connections

stavros11 committed:
feat: add debug as an option

stavros11 committed:
fix: fit and report when save_sequences is False

stavros11 committed:
fix: typo in fit results

stavros11 committed:
feat: option to run on logarithmic scale depths

stavros11 commented on src/qibocal/protocols/qua/rb_ondevice.py:
```diff
@@ -0,0 +1,565 @@
+
+            if save_sequences:
+                save(sequence_list[0], sequence_st)  # save sequence indices in stream
+            with for_(
+                depth, 1, depth <= max_circuit_depth, depth + 1
+            ):  # Loop over the depths
```
I think this loop needs to have step 1 in every cases, because a few lines below we are saving `sequence_list[depth]` to the stream that is then used to return the sequence indices to the user if `save_sequences=True`. And we need to save the indices at all depths, even when we are not running on every depth (regardless of logarithmic or no). Therefore, `for_each_` wouldn't be useful here.

alecandido replied:
Couldn't you use another variable to index it?

The equivalent of `for i, depth in enumerate(depths)` and then save in `sequence_list[i]`.
(of course QUA doesn't have `enumerate()` on hardware, but you could realize just by initializing a counter and increment by 1 at every loop)

stavros11 commented:
@ingoroth in a74c0ee370ca4d7aa8ab00008b67c42fae208097 I added the possibility to increase depths logarithmically. It can be enabled using `logarithmic: True` and `delta_clifford` will then act as the exponent, so the depths should be 2, 2^delta_clifford, 2^(2*delta_clifford), ..., up to `max_circuit_depth`. An example action runcard is:
```yaml
  - id: RB QUA
    operation: rb_ondevice
    parameters:
      num_of_sequences: 10000  # Number of random sequences.
      max_circuit_depth: 1024 # Maximum circuit depth.
      delta_clifford: 1
      logarithmic: True
      n_avg: 1  # Number of repetitions of the same circuit (for averaging).
      save_sequences: True  # Download the Clifford gate indices used from the device.
      apply_inverse: True  # Apply inverse gate at the end of every circuit. If `False` it will apply "filtered" post-processing.
```
I did a quick test and it seems to work (and is quite faster) up to depth 1024, but failed for 2048.

stavros11 committed:
feat: import qua-libs code for 2q RB

stavros11 committed:
fix: element names

stavros11 committed:
feat: read CZ parameters from platform

stavros11 committed:
fix: results

stavros11 committed:
chore: drop test directory

stavros11 committed:
fix: pylint

stavros11 committed:
fix: run on hardware

stavros11 committed:
feat: params.debug option to dump QUA script

stavros11 committed:
feat: allow different order of qubits

stavros11 committed:
chore: disable initial fitting guess

stavros11 committed:
chore: drop prints

stavros11 committed:
chore: compatibility with OPX1000

stavros11 commented:
This was not compatible with the OPX1000 due to some bugs related to the latest qm-qua prerelease or its firmware, in particular:
1. Voltage offsets (used for sweetspots) are not set properly from the config, so we need to force the using the `set_dc_offset` in the QUA program (this is a bug in QOP320 confirmed by QM).
2. Strict timing throws an error which was not happening in the OPX+. I asked QM to check this and temporarily bypassed this using compiler options.

The workarounds for these were done in 7d22c7b86dc5003f2f3fb2c135a6756eb56c4624 and now this should work also for the OPX1000, see for example a report with the qw5q_platinum: http://login.qrccluster.com:9000/KjIpqUWCQw6-ehT9HlgvTw==

@biankawolo @igres26 you may be interested in this. If you attempt to pull, please use `git pull --rebase` because I also rebased this to the latest main.

stavros11 committed:
chore: compatibility with OPX1000

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stavros11 committed:
feat: add flux waveforms in config (needed for 2q RB)

stavros11 committed:
fix: drop redundant line

stavros11 committed:
fix: drop redundant check

stavros11 committed:
refactor: rename params.debug to params.script_file

stavros11 committed:
refactor: open QM after dumping script

stavros11 committed:
fix: drop duplicate method

stavros11 committed:
feat: allow running RB on multiple qubits at the same time

stavros11 committed:
feat: implement 2q RB using Qiskit circuits

stavros11 committed:
fix: add delay only when >0

stavros11 committed:
refactor: standardize routine names

stavros11 committed:
chore: update for qibolab 0.2

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stavros11 committed:
fix: execute single qubit RB on 0.2

stavros11 committed:
fix: CZ does not start with flux pulse case

stavros11 committed:
feat: allow arbitrary waveforms for flux pulse and measurement

stavros11 committed:
fix: 2q rb executes

stavros11 commented:
Since running fast RB on QM is still useful, I updated this to be compatible with qibolab 0.2 platforms. The old 0.1 version is still available in a different branch: [`qua-0.1`](https://github.com/qiboteam/qibocal/tree/qua-0.1). Both branches include the single and two-qubit standard RB from qua-libs:
1. https://github.com/qua-platform/qua-libs/blob/main/Quantum-Control-Applications/Superconducting/Single-Flux-Tunable-Transmon/16a_randomized_benchmarking.py
2. https://github.com/qua-platform/qua-libs/tree/main/Quantum-Control-Applications/Superconducting/Two-Flux-Tunable-Coupled-Transmons/Use%20Case%202%20-%20Two-Qubit-Randomized-Benchmarking

I also renamed the routines to be more consistent, so here are the example parameters:
```yaml
- id: QUA Standard RB
    operation: qua_standard_rb_1q
    parameters:
      num_of_sequences: 8000
      max_circuit_depth: 300
      delta_clifford: 10
      # logarithmic: True
      n_avg: 1
      save_sequences: False
      apply_inverse: True
```
and
```yaml
id: QUA Standard two-qubit RB
  operation: qua_standard_rb_2q
  parameters:
    circuit_depths: [1, 2, 3, 4, 5]
    num_circuits_per_depth: 50
    num_shots_per_circuit: 100
```

stavros11 committed:
fix: baked pulse duration

stavros11 committed:
feat: support different sampling rates

stavros11 committed:
feat: add qua-libs cryoscope routine

stavros11 committed:
feat: fast qiskit RB in qua

stavros11 committed:
fix: flux pulse amplified amplitude

stavros11 committed:
fix: temporary fix for qubit ordering

Luca-Ben-Herrmann committed:
Fix fidelity caluculation

stavros11 committed:
feat: interleave_cz option

stavros11 committed:
feat: qua state and process tomography

stavros11 committed:
fix: imports

stavros11 committed:
feat: implement batch size in stream RB

stavros11 committed:
fix: rebase

stavros11 committed:
fix: imports

stavros11 committed:
feat: read CZ sequence from runcard

stavros11 committed:
refactor: drop debug option and dump qua script in path specified in the platform

stavros11 committed:
fix: ignore channels that have no config

stavros11 committed:
fix: support readout with probe.duration != acquisition.duration in configuration

stavros11 committed:
fix: add qua protocols to __init__
